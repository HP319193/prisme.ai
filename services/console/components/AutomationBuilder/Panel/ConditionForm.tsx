import { useTranslation } from 'next-i18next';
import { Button } from 'primereact/button';
import { FC, useCallback } from 'react';
import { Form } from 'react-final-form';
import Fieldset from '../../../layouts/Fieldset';
import { Field } from '../../SchemaForm/Field';

interface ConditionFormProps {
  condition?: string;
  onSubmit: (c: string) => void;
}

export const ConditionForm: FC<ConditionFormProps> = ({
  condition,
  onSubmit,
}) => {
  const { t } = useTranslation('workspaces');
  const submit = useCallback(
    (values) => {
      onSubmit(values.condition);
    },
    [onSubmit]
  );

  return (
    <div>
      <Form onSubmit={submit} initialValues={{ condition }}>
        {({ handleSubmit }) => (
          <form onSubmit={handleSubmit}>
            <Fieldset legend={t('automations.condition.edit.title')}>
              <Field field="condition" type="string" required />
              <Button type="submit">
                {t('automations.condition.edit.submit')}
              </Button>
            </Fieldset>
          </form>
        )}
      </Form>
    </div>
  );
};

export default ConditionForm;
