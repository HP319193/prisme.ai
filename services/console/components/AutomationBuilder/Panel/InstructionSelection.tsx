import { ListItem, SearchInput, Space, Title } from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';
import Image from 'next/image';
import { Button } from '@prisme.ai/design-system';
import { FC, useMemo, useState } from 'react';
import { useAutomationBuilder } from '../context';

export interface InstructionSelectionProps {
  onSubmit: (key: string) => void;
  focus?: true;
}

export const InstructionSelection: FC<InstructionSelectionProps> = ({
  onSubmit,
  focus,
}) => {
  const { t } = useTranslation('workspaces');
  const { instructionsSchemas } = useAutomationBuilder();

  const [search, setSearch] = useState('');
  const filteredInstructions = useMemo(() => {
    return instructionsSchemas
      .reduce<[string, string, { name: string; description?: string }[]][]>(
        (prev, [name, list, more]) => {
          const matching = (
            search
              ? Object.keys(list).filter((a) =>
                  `${name} ${a}`.toLowerCase().match(search.toLowerCase())
                )
              : Object.keys(list)
          ).map((name) => ({
            name,
            description: (list[name] || {}).description,
          }));
          if (!matching) return prev;
          return [...prev, [name, more.icon, matching]];
        },
        []
      )
      .filter(([, , list]) => list.length);
  }, [instructionsSchemas, search]);

  return (
    <div className="flex grow h-full flex-col overflow-auto">
      <SearchInput
        value={search}
        onChange={(e) => setSearch(e.target.value)}
        placeholder={t('automations.instruction.search')}
        autoFocus={focus}
        className="mb-6"
      />
      <Space direction="vertical" className="flex grow overflow-x-auto">
        {filteredInstructions.map(([section, icon, instructions]) => (
          <Space key={section} direction="vertical" className="!flex flex-1">
            <Space>
              <Image src={icon} width={16} height={16} alt={section} />
              <Title level={4}>{section}</Title>
            </Space>
            <Space direction="vertical" className="!flex flex-1">
              {instructions.map(({ name, description }) => (
                <Button
                  key={name}
                  onClick={() => onSubmit(name)}
                  className="w-full text-left !h-fit"
                >
                  <ListItem title={name} /*content={description}*/ />
                </Button>
              ))}
            </Space>
          </Space>
        ))}
      </Space>
    </div>
  );
};

export default InstructionSelection;
