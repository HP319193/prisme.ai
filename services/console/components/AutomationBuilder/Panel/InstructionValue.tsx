import { FC, useCallback, useMemo, useState } from 'react';
import Fieldset from '../../../layouts/Fieldset';
import { useTranslation } from 'next-i18next';
import {
  FieldProps,
  getSchemaFormLabel,
  SchemaForm,
  SchemaFormDescription,
} from '@prisme.ai/design-system';
import { CodeEditorInline } from '../../CodeEditor/lazy';
import { useField } from 'react-final-form';
import FieldContainerWithRaw from '../../FieldContainerWithRaw';
import useSchema from '../../SchemaForm/useSchema';
import { useWorkspace } from '../../../layouts/WorkspaceLayout';
import usePages from '../../PagesProvider/context';

interface InstructionValueProps {
  instruction: string;
  value: any;
  schema?: any;
  onChange: (values: any) => void;
}

const EmptyButtons: any[] = [];
const FieldAny = ({ schema, name, label }: FieldProps) => {
  const field = useField(name);
  const [value, setValue] = useState(
    typeof field.input.value === 'string'
      ? field.input.value
      : JSON.stringify(field.input.value, null, '  ')
  );
  const onChange = useCallback(
    (value: string) => {
      setValue(value);
      try {
        const json = JSON.parse(value);
        field.input.onChange(json);
      } catch {
        field.input.onChange(value);
      }
    },
    [field.input]
  );

  return (
    <div className="flex flex-1 flex-col my-2">
      <SchemaFormDescription text={schema.description}>
        <label className="text-[10px] text-gray">
          {label || schema.title || getSchemaFormLabel(name)}
        </label>
        <CodeEditorInline
          value={value}
          onChange={onChange}
          mode="json"
          style={{ flex: 'auto' }}
        />
      </SchemaFormDescription>
    </div>
  );
};

const components = {
  FieldAny,
  FieldContainer: FieldContainerWithRaw,
};

export const InstructionValue: FC<InstructionValueProps> = ({
  instruction,
  value,
  schema = {},
  onChange,
}) => {
  const { workspace } = useWorkspace();
  const { pages } = usePages();
  const { t } = useTranslation('workspaces');

  const appInstance = useMemo(() => {
    if (!workspace.imports) return workspace.config;
    const [appName] = instruction.split(/\./);
    if (!workspace.imports[appName]) return workspace.config;
    return workspace.imports[appName].config || {};
  }, [instruction, workspace.config, workspace.imports]);

  const { extractSelectOptions } = useSchema({
    config: appInstance,
    automations: workspace.automations,
    pages: pages.get(workspace.id),
  });

  const cleanedSchema = useMemo(() => {
    const cleaned = {
      ...schema,
      title: t('automations.instruction.description', {
        context: instruction,
        default: schema.description,
      }),
    };
    if (instruction === 'repeat') {
      cleaned['ui:options'] = {
        oneOf: {
          options: [
            {
              label: 'automations.instruction.form.repeat.on.label',
              index: 0,
              value: {
                until: undefined,
              },
            },
            {
              label: 'automations.instruction.form.repeat.until.label',
              index: 1,
              value: {
                on: undefined,
              },
            },
          ],
        },
      };
    }
    return cleaned;
  }, [instruction, schema, t]);

  const locales = useMemo(
    () => ({
      addItem: t('automations.instruction.form.addItem'),
      addProperty: t('automations.instruction.form.addProperty'),
      propertyKey: t('automations.instruction.form.propertyKey'),
      propertyValue: t('automations.instruction.form.propertyValue'),
      removeItem: t('automations.instruction.form.removeItem'),
      removeProperty: t('automations.instruction.form.removeProperty'),
      uploadLabel: t('automations.instruction.form.uploadLabel'),
      uploadRemove: t('automations.instruction.form.uploadRemove'),
    }),
    [t]
  );

  if (!schema) return null;

  return (
    <Fieldset
      legend={t('automations.instruction.label', { context: instruction })}
      hasDivider={false}
    >
      <SchemaForm
        schema={cleanedSchema}
        onChange={onChange}
        initialValues={value}
        buttons={EmptyButtons}
        components={components}
        locales={locales}
        utils={{ extractSelectOptions }}
      />
    </Fieldset>
  );
};

export default InstructionValue;
