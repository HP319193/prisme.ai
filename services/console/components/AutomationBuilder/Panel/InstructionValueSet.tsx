import {
  FieldProps,
  InfoBubble,
  SchemaForm,
  SchemaFormProps,
} from '@prisme.ai/design-system';
import { useEffect, useMemo } from 'react';
import Selector from '../../SchemaForm/BlockSelector/Selector';
import {
  BlockSelectorProvider,
  useBlockSelector,
} from '../../SchemaForm/BlockSelector/BlockSelectorProvider';
import { useField } from 'react-final-form';
import { CloseCircleOutlined } from '@ant-design/icons';
import components, { FieldAny } from '../../SchemaForm/schemaFormComponents';
import { useTranslation } from 'next-i18next';
import FieldContainerWithRaw from '../../FieldContainerWithRaw';

export interface InstructionValueSetProps extends SchemaFormProps {}

const InterfaceSelector = (props: FieldProps) => {
  const { selectBlock } = useBlockSelector();
  const { t } = useTranslation('workspaces');
  const field = useField(props.name);

  useEffect(() => {
    selectBlock(field.input.value);
  }, [field.input.value, selectBlock]);

  return (
    <div className="pr-form-field">
      <label htmlFor={`${field.input.name}.name`} className="pr-form-label">
        {t('automations.instruction.form.set.interface.label')}
      </label>
      <div className="pr-form-input relative">
        <Selector
          id={`${field.input.name}.name`}
          {...field.input}
          className="w-full"
        />
        <button
          onClick={() => field.input.onChange(null)}
          className="absolute top-2 right-8"
        >
          <CloseCircleOutlined />
        </button>
      </div>
      <div className="pr-form-description">
        <InfoBubble
          className="pr-form-object__description"
          text={props.schema.description}
        />
      </div>
    </div>
  );
};

const ValueEditor = (props: FieldProps) => {
  const { t } = useTranslation('workspaces');
  const { schema } = useBlockSelector();
  const field = useField(props.name);

  if (schema)
    return (
      <FieldContainerWithRaw {...props}>
        <label htmlFor={`${field.input.name}.name`} className="pr-form-label">
          {t('automations.instruction.form.set.value.label')}
        </label>
        <div className="pr-form-input flex flex-1 -m-4 mt-4">
          <SchemaForm
            schema={schema}
            initialValues={field.input.value}
            onChange={field.input.onChange}
            components={components}
          />
        </div>
        <div className="pr-form-description">
          <InfoBubble text={props.schema.description} />
        </div>
      </FieldContainerWithRaw>
    );
  return <FieldAny {...props} />;
};

export const InstructionValueSet = (props: InstructionValueSetProps) => {
  const schema = useMemo(() => {
    const schema = props.schema;
    schema.properties = schema.properties || {};
    schema.properties.interface['ui:widget'] = InterfaceSelector;

    schema.properties.value['ui:widget'] = ValueEditor;

    return schema;
  }, [props.schema]);

  return (
    <BlockSelectorProvider>
      <SchemaForm {...props} schema={schema} />
    </BlockSelectorProvider>
  );
};
