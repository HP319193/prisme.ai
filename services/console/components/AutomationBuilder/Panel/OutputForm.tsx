import { FC, useMemo } from 'react';
import { Schema, SchemaForm } from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';
import components from './schemaFormComponents';

interface OutputFormProps {
  output?: string;
  onChange: (v: { output: any }) => void;
}

const buttons: any[] = [];

export const OutputForm: FC<OutputFormProps> = ({ output, onChange }) => {
  const { t } = useTranslation('workspaces');

  const schema: Schema = useMemo(
    () => ({
      additionalProperties: true,
      title: t('automations.output.edit.title'),
      description: t('automations.output.edit.description', {
        interpolation: {
          skipOnVariables: true,
        },
      }),
    }),
    [t]
  );

  return (
    <div className="flex flex-1 flex-col h-full overflow-x-auto">
      <SchemaForm
        schema={schema}
        onChange={onChange}
        initialValues={output}
        components={components}
        buttons={buttons}
      />
    </div>
  );
};

export default OutputForm;
