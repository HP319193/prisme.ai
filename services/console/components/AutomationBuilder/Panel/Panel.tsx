import { FC, useRef } from 'react';
import { Sidebar } from 'primereact/sidebar';

interface PanelProps {
  visible: boolean;
  onVisibleChange?: (v: boolean) => void;
}
export const Panel: FC<PanelProps> = ({
  visible,
  onVisibleChange,
  children,
}) => {
  const ref = useRef(null);

  return (
    <>
      <div ref={ref} className="absolute top-0 bottom-0 right-0" />
      {ref.current && (
        <Sidebar
          visible={visible}
          onHide={() => onVisibleChange && onVisibleChange(false)}
          position="right"
          modal={false}
          appendTo={ref.current}
          maskStyle={{ position: 'relative' }}
          className="w-30rem"
        >
          {children}
        </Sidebar>
      )}
    </>
  );
};

export default Panel;
