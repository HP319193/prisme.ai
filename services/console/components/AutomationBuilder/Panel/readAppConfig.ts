import { JSONPath } from 'jsonpath-plus';

export const readAppConfig = (appConfig: any, path: string) => {
  return JSONPath({ path, json: appConfig });
};
