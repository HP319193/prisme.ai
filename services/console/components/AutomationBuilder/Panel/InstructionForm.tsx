import { FC, useCallback, useEffect, useState } from 'react';
import InstructionSelection from './InstructionSelection';
import InstructionValue from './InstructionValue';
import { Schema } from '../../SchemaForm/types';
import { useAutomationBuilder } from '../context';
import { ArrowLeftOutlined } from '@ant-design/icons';

const getDefaultValue = (type: string) => {
  switch (type) {
    case 'array':
      return [];
    case 'object':
    default:
      return {};
  }
};
interface InstructionFormProps {
  instruction?: Prismeai.Instruction;
  onSubmit: (i: Prismeai.Instruction) => void;
}

export const InstructionForm: FC<InstructionFormProps> = ({
  instruction,
  onSubmit,
}) => {
  const { getSchema } = useAutomationBuilder();
  const [currentInstruction] = Object.keys(instruction || {});
  const [edit, setEdit] = useState<{
    instruction: string;
    value: any;
    schema: Schema | null;
  }>({
    instruction: currentInstruction,
    value:
      currentInstruction &&
      instruction &&
      instruction[currentInstruction as keyof typeof instruction],
    schema: getSchema(currentInstruction),
  });

  const setInstruction = useCallback(
    (instructionName: string) => {
      const schema = getSchema(instructionName);
      if (!schema.properties) {
        onSubmit({
          [instructionName]: getDefaultValue(schema.type || ''),
        });
        return;
      }
      setEdit({
        instruction: instructionName,
        value: {},
        schema,
      });
    },
    [getSchema, onSubmit]
  );

  useEffect(() => {
    const [currentInstruction] = Object.keys(instruction || {});
    setEdit({
      instruction: currentInstruction,
      value:
        currentInstruction &&
        instruction &&
        instruction[currentInstruction as keyof typeof instruction],
      schema: getSchema(currentInstruction),
    });
  }, [getSchema, instruction, setInstruction]);

  const setInstructionValue = useCallback(
    (values: any) => {
      if (!edit.instruction) return;
      onSubmit({
        [edit.instruction]: values,
      });
    },
    [edit.instruction, onSubmit]
  );

  const unsetInstruction = useCallback(() => {
    setEdit({
      instruction: '',
      value: {},
      schema: null,
    });
  }, []);

  return (
    <div className="flex flex-1 flex-col h-full overflow-x-auto">
      {!instruction && edit.instruction && (
        <button
          onClick={unsetInstruction}
          className="absolute top-[24px] left-[1rem]"
        >
          <ArrowLeftOutlined />
        </button>
      )}
      {!edit ||
        (!edit.instruction && (
          <InstructionSelection onSubmit={setInstruction} />
        ))}
      {edit && edit.instruction && (
        <InstructionValue
          instruction={edit.instruction}
          value={edit.value}
          schema={edit.schema}
          onSubmit={setInstructionValue}
        />
      )}
    </div>
  );
};

export default InstructionForm;
