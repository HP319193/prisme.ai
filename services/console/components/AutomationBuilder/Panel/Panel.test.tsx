import Panel from './Panel';
import renderer, { act } from 'react-test-renderer';
import { Sidebar } from 'primereact/sidebar';

jest.mock('primereact/sidebar', () => {
  const Sidebar = () => <div>Sidebar</div>;
  return {
    Sidebar,
  };
});

it('should render', () => {
  const onVisibleChange = jest.fn();
  const div = {};
  const root = renderer.create(
    <Panel visible={true} onVisibleChange={onVisibleChange} />,
    {
      createNodeMock(el) {
        if (el.type === 'div') return div;
        return null;
      },
    }
  );
  expect(root.toJSON()).toMatchSnapshot();

  act(() => {
    root.update(<Panel visible={true} onVisibleChange={onVisibleChange} />);
  });

  expect(root.toJSON()).toMatchSnapshot();

  root.root.findByType(Sidebar).props.onHide();

  expect(onVisibleChange).toHaveBeenCalledWith(false);
});
