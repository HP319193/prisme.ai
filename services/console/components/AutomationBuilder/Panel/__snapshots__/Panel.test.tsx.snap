// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should hide 1`] = `
<div
  className="
        flex
        absolute top-0 bottom-0 -right-1/3 w-1/3 z-10 flex-col
        transition-transform
        ease-in
        duration-200
        overflow-hidden
        
      "
>
  <div
    className="border border-gray-200 border-solid grow h-full rounded p-4"
  >
    <div
      className="flex flex-col overflow-hidden h-full pb-12"
    >
      <div
        className="flex justify-end"
      />
    </div>
  </div>
</div>
`;

exports[`should render 1`] = `
<div
  className="
        flex
        absolute top-0 bottom-0 -right-1/3 w-1/3 z-10 flex-col
        transition-transform
        ease-in
        duration-200
        overflow-hidden
        
      "
>
  <div
    className="border border-gray-200 border-solid grow h-full rounded p-4"
  >
    <div
      className="flex flex-col overflow-hidden h-full pb-12"
    >
      <div
        className="flex justify-end"
      />
    </div>
  </div>
</div>
`;
